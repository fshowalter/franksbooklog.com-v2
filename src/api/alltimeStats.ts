import type { AlltimeStatsJson } from "./data/alltimeStatsJson";
import { alltimeStatsJson } from "./data/alltimeStatsJson";

export interface AlltimeStats extends AlltimeStatsJson {}

export async function alltimeStats(): Promise<AlltimeStats> {
  return await alltimeStatsJson();
}
