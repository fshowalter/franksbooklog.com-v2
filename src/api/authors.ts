import { allAuthorsJson, type AuthorJson } from "./data/authorsJson";

export interface Author extends AuthorJson {}

export async function allAuthors(): Promise<Author[]> {
  return await allAuthorsJson();
}
