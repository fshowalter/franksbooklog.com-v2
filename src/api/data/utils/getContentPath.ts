import { join } from "path";

export function getContentPath(
  kind: "data" | "reviews" | "readings" | "pages",
  path?: string,
) {
  if (path) {
    return join(process.cwd(), "content", kind, path);
  }

  return join(process.cwd(), "content", kind);
}
