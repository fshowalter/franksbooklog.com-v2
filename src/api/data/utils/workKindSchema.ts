import { z } from "zod";

export const WorkKindSchema = z.enum([
  "Anthology",
  "Collection",
  "Nonfiction",
  "Novel",
  "Novella",
  "Short Story",
]);
