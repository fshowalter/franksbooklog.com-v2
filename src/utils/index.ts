export * from "./buildGroupValues";
export * from "./filterTools";
export * from "./sortTools";
export * from "./toSentenceArray";
