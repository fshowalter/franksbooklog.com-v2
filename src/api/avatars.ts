import { getImage } from "astro:assets";

import { normalizeSources } from "./utils/normalizeSources";

export interface AvatarImageData {
  src: string;
  srcSet: string;
}

interface Props {
  width: number;
  height: number;
  authors: { slug: string }[];
}

const images = import.meta.glob<{ default: ImageMetadata }>(
  "/content/assets/avatars/*.png",
);

export async function getAvatars({
  authors,
  width,
  height,
}: Props): Promise<Record<string, AvatarImageData>> {
  const imageMap: Record<string, AvatarImageData> = {};

  await Promise.all(
    authors.map(async (author) => {
      const imagePath = Object.keys(images).find((path) => {
        return path.endsWith(`${author.slug}.png`);
      });

      if (!imagePath) {
        return null;
      }

      const avatarFile = await images[imagePath]();

      const optimizedImage = await getImage({
        src: avatarFile.default,
        width: width,
        height: height,
        format: "avif",
        densities: [1, 2],
        quality: 80,
      });

      imageMap[author.slug] = {
        srcSet: normalizeSources(optimizedImage.srcSet.attribute),
        src: normalizeSources(optimizedImage.src),
      };
    }),
  );

  return imageMap;
}
