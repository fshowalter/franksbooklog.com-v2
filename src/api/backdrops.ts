import { getImage } from "astro:assets";

import { normalizeSources } from "./utils/normalizeSources";

export interface BackdropImageData {
  src: string;
  srcSet: string;
}

interface Props {
  slug: string;
  width: number;
  height: number;
}

export const images = import.meta.glob<{ default: ImageMetadata }>(
  "/content/assets/backdrops/*.png",
);

export async function getBackdrop({
  slug,
  width,
  height,
}: Props): Promise<BackdropImageData> {
  const backdropPath = Object.keys(images).find((path) => {
    return path.endsWith(`${slug}.png`);
  })!;

  const backdropFile = await images[backdropPath]();

  const optimizedImage = await getImage({
    src: backdropFile.default,
    width: width,
    height: height,
    format: "avif",
    widths: [0.25, 0.5, 1, 2].map((w) => w * width),
    quality: 80,
  });

  return {
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    src: normalizeSources(optimizedImage.src),
  };
}
