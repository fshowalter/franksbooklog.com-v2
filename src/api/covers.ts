import { getImage } from "astro:assets";

import { normalizeSources } from "./utils/normalizeSources";

export interface CoverImageData {
  src: string;
  srcSet: string;
}

interface Work {
  slug: string;
  includedInSlugs: string[];
}

interface Props {
  works: Work[];
  width: number;
  height: number;
}

const images = import.meta.glob<{ default: ImageMetadata }>(
  "/content/assets/covers/*.png",
);

function parentCoverForWork(work: Work) {
  let parentCover;

  work.includedInSlugs.find((slug) => {
    parentCover = Object.keys(images).find((image) => {
      return image.endsWith(`${slug}.png`);
    });

    return parentCover;
  });

  return parentCover;
}

export async function getCovers({
  works,
  width,
  height,
}: Props): Promise<Record<string, CoverImageData>> {
  const imageMap: Record<string, CoverImageData> = {};

  await Promise.all(
    works.map(async (work) => {
      let workCoverPath = Object.keys(images).find((image) => {
        return image.endsWith(`${work.slug}.png`);
      });

      if (!workCoverPath) {
        workCoverPath = parentCoverForWork(work)!;
      }

      const workCoverFile = await images[workCoverPath]();

      const optimizedImage = await getImage({
        src: workCoverFile.default,
        width: width,
        height: height,
        format: "avif",
        densities: [1, 2],
        quality: 80,
      });

      imageMap[work.slug] = {
        srcSet: normalizeSources(optimizedImage.srcSet.attribute),
        src: normalizeSources(optimizedImage.src),
      };
    }),
  );

  return imageMap;
}
