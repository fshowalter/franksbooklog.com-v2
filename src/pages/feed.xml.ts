import rss from "@astrojs/rss";
import { getFeedCover } from "src/api/covers";
import {
  loadExcerptHtml,
  mostRecentReviews,
  type ReviewWithExcerpt,
} from "src/api/reviews";
import { textStarsForGrade } from "src/utils/textStarsForGrade";

function addMetaToExcerpt(excerpt: string, review: ReviewWithExcerpt) {
  const meta = `${textStarsForGrade(review.grade)}`;
  return `<p>${meta}</p>${excerpt}`;
}

function authorsToString(authors: ReviewWithExcerpt["authors"]) {
  const authorsArray = authors.map((author) => {
    if (author.notes) {
      return `${author.name} (${author.notes})`;
    }

    return author.name;
  });

  return new Intl.ListFormat().format(authorsArray);
}

export async function GET() {
  const reviews = await mostRecentReviews(10);

  const rssItems = await Promise.all(
    reviews.map(async (review) => {
      return await loadExcerptHtml(review);
    }),
  );

  return rss({
    // `<title>` field in output xml
    title: "Frank's Book Log",
    // `<description>` field in output xml
    description: "Reviews of current, cult, classic, and forgotten books.",
    // Pull in your project "site" from the endpoint context
    // https://docs.astro.build/en/reference/api-reference/#contextsite
    site: "https://www.franksbooklog.com",
    customData:
      "<image><url>https://www.franksbooklog.com/assets/favicon-128.png</url><title>Frank's Book Log</title><link>https://www.franksbooklog.com/</link></image>",
    // Array of `<item>`s in output xml
    // See "Generating items" section for examples using content collections and glob imports
    items: await Promise.all(
      rssItems.map(async (item) => {
        const cover = await getFeedCover(item);

        return {
          title: `${item.title} by ${authorsToString(item.authors)}`,
          pubDate: item.date,
          link: `https://www.franksbooklog.com/reviews/${item.slug}/`,
          content: `<img src="${
            cover.src
          }" alt="${cover.alt}">${addMetaToExcerpt(item.excerpt, item)}`,
        };
      }),
    ),
  });
}
