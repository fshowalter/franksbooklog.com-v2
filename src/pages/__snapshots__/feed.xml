<?xml version="1.0" encoding="UTF-8"?><rss version="2.0" xmlns:content="http://purl.org/rss/1.0/modules/content/"><channel><title>Frank&apos;s Book Log</title><description>Reviews of current, cult, classic, and forgotten books.</description><link>https://www.franksbooklog.com/</link><image><url>https://www.franksbooklog.com/assets/favicon-128.png</url><title>Frank&apos;s Book Log</title><link>https://www.franksbooklog.com/</link></image><item><title>Fast Forward by Stephen Morris</title><link>https://www.franksbooklog.com/reviews/fast-forward-by-stephen-morris/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/fast-forward-by-stephen-morris/</guid><pubDate>Mon, 01 Jul 2024 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Ffast-forward-by-stephen-morris.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for Fast Forward: Confessions of a Post-Punk Percussionist: Volume II by Stephen Morris&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#189;&lt;/p&gt;&lt;p&gt;Every bit as entertaining as his first volume, drummer Stephen Morris’s second volume of memoirs picks up where &lt;em&gt;Record Play Pause&lt;/em&gt; ended, chronicling New Order’s rise to super-stardom and the tensions that tore it apart. &lt;a class=&quot;!no-underline uppercase whitespace-nowrap text-accent text-sm leading-none&quot; href=&quot;/reviews/fast-forward-by-stephen-morris/&quot;&gt;Continue reading...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item><title>Gothic by Philip Fracassi</title><link>https://www.franksbooklog.com/reviews/gothic-by-philip-fracassi/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/gothic-by-philip-fracassi/</guid><pubDate>Sun, 30 Jun 2024 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Fgothic-by-philip-fracassi.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for Gothic by Philip Fracassi&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;blockquote&gt;
&lt;p&gt;“It’s like Christine… but wood!”&lt;/p&gt;
&lt;/blockquote&gt;
&lt;p&gt;So remarks a character in Philip Fracassi’s meta horror novel about a haunted desk. And while the statement is true per se, it doesn’t do justice to Fracassi’s work, which proves far leaner and meaner than King’s. &lt;a class=&quot;!no-underline uppercase whitespace-nowrap text-accent text-sm leading-none&quot; href=&quot;/reviews/gothic-by-philip-fracassi/&quot;&gt;Continue reading...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item><title>Record Play Pause by Stephen Morris</title><link>https://www.franksbooklog.com/reviews/record-play-pause-by-stephen-morris/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/record-play-pause-by-stephen-morris/</guid><pubDate>Sat, 11 Nov 2023 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Frecord-play-pause-by-stephen-morris.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for Record Play Pause by Stephen Morris&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;Warm, witty, self-deprecating and self-aware, drummer Stephen Morris’s first volume of memoirs chronicles his formative years in the Manchester suburb of Macclesfield through joining Joy Division, its end, and the early days of New Order. &lt;a class=&quot;!no-underline uppercase whitespace-nowrap text-accent text-sm leading-none&quot; href=&quot;/reviews/record-play-pause-by-stephen-morris/&quot;&gt;Continue reading...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item><title>Fright Night Origins by Tom Holland and A. Jack Ulrich</title><link>https://www.franksbooklog.com/reviews/fright-night-origins-by-tom-holland-a-jack-ulrich/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/fright-night-origins-by-tom-holland-a-jack-ulrich/</guid><pubDate>Mon, 06 Nov 2023 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Ffright-night-origins-by-tom-holland-a-jack-ulrich.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for Fright Night Origins by Tom Holland and A. Jack Ulrich&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;Tom Holland, director and screenwriter of the original &lt;em&gt;Fright Night&lt;/em&gt; film, partners with A. Jack Ulrich to craft an expanded novelization of the film, proffering more backstory and setting up an expanded universe and sequel. &lt;a class=&quot;!no-underline uppercase whitespace-nowrap text-accent text-sm leading-none&quot; href=&quot;/reviews/fright-night-origins-by-tom-holland-a-jack-ulrich/&quot;&gt;Continue reading...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item><title>Don&apos;t Let Them Get You Down by Philip Fracassi</title><link>https://www.franksbooklog.com/reviews/dont-let-them-get-you-down-by-philip-fracassi/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/dont-let-them-get-you-down-by-philip-fracassi/</guid><pubDate>Sun, 29 Oct 2023 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Fdont-let-them-get-you-down-by-philip-fracassi.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for Don&apos;t Let Them Get You Down by Philip Fracassi&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#189;&lt;/p&gt;&lt;p&gt;&lt;em&gt;Don’t Let Them Get You Down&lt;/em&gt; follows Peter, a copywriter for a medical website, struggling to hold his life together while battling a spiraling depression triggered by his wife’s cancer death two years prior. &lt;a class=&quot;!no-underline uppercase whitespace-nowrap text-accent text-sm leading-none&quot; href=&quot;/reviews/dont-let-them-get-you-down-by-philip-fracassi/&quot;&gt;Continue reading...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item><title>The Creative Act by Rick Rubin</title><link>https://www.franksbooklog.com/reviews/the-creative-act-by-rick-rubin/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/the-creative-act-by-rick-rubin/</guid><pubDate>Sun, 22 Oct 2023 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Fthe-creative-act-by-rick-rubin.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for The Creative Act: A Way of Being by Rick Rubin&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;Rick Rubin, the famed music producer of such disparate acts as the Beastie Boys, Slayer, Tom Petty, and Johnny Cash, proffers a meditation on creativity. &lt;a class=&quot;!no-underline uppercase whitespace-nowrap text-accent text-sm leading-none&quot; href=&quot;/reviews/the-creative-act-by-rick-rubin/&quot;&gt;Continue reading...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item><title>Sacculina by Philip Fracassi</title><link>https://www.franksbooklog.com/reviews/sacculina-by-philip-fracassi/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/sacculina-by-philip-fracassi/</guid><pubDate>Mon, 09 Oct 2023 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Fsacculina-by-philip-fracassi.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for Sacculina by Philip Fracassi&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;To celebrate Jim’s older brother Jack’s release following a six-year prison stint, the brothers, their father Henry, and Jack’s hulking best friend Chris, charter a fishing boat in the Pacific. The choppy seas don’t bode well for an outing, but the group presses ahead, and their grizzled captain takes them a few miles offshore to a deep spot good for fishing. But the men soon realize they’re under threat by an unknown, overwhelming assailant—and how alone they are at sea. &lt;a class=&quot;!no-underline uppercase whitespace-nowrap text-accent text-sm leading-none&quot; href=&quot;/reviews/sacculina-by-philip-fracassi/&quot;&gt;Continue reading...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item><title>The Cipher by Kathe Koja</title><link>https://www.franksbooklog.com/reviews/the-cipher-by-kathe-koja/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/the-cipher-by-kathe-koja/</guid><pubDate>Sun, 29 Jan 2023 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Fthe-cipher-by-kathe-koja.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for The Cipher by Kathe Koja&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;How to describe &lt;em&gt;The Cipher&lt;/em&gt;? The plot concerns Nicholas, an aimless twenty-something drifting through life, who discovers a mysterious hole inside a disused storage room in his derelict apartment building. Weird things ensue. &lt;a class=&quot;!no-underline uppercase whitespace-nowrap text-accent text-sm leading-none&quot; href=&quot;/reviews/the-cipher-by-kathe-koja/&quot;&gt;Continue reading...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item><title>Roadside Pickup by Richard Laymon</title><link>https://www.franksbooklog.com/reviews/roadside-pickup-by-richard-laymon/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/roadside-pickup-by-richard-laymon/</guid><pubDate>Sat, 21 Jan 2023 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Fshort-stories-volume-one-by-richard-laymon.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for Short Stories: Volume One: The Mystery and Men&apos;s Magazines by Richard Laymon&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#189;&lt;/p&gt;&lt;p&gt;Early morning darkness. Colleen sits stranded in her car. Three hours have produced three cars. None have stopped. But the fourth does. A young man with fashionable clothes and booze on his breath. Won’t she let him give her a ride?&lt;/p&gt;
&lt;p&gt;Laymon’s second published story proffers a predictable twist, but at under two thousand words, it doesn’t drag.&lt;/p&gt;</content:encoded></item><item><title>Fright Night by John Skipp and Craig Spector</title><link>https://www.franksbooklog.com/reviews/fright-night-by-john-skipp-craig-spector/</link><guid isPermaLink="true">https://www.franksbooklog.com/reviews/fright-night-by-john-skipp-craig-spector/</guid><pubDate>Sat, 21 Jan 2023 00:00:00 GMT</pubDate><content:encoded>&lt;img src=&quot;/_image?/content%2Fassets%2Fcovers%2Ffright-night-by-john-skipp-craig-spector.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;A cover for Fright Night by John Skipp and Craig Spector&quot;&gt;&lt;p&gt;&amp;#9733;&lt;/p&gt;&lt;p&gt;During a late-night make-out session with his girlfriend, high-schooler Charlie Brewster spies two men carrying a coffin into the basement of the house next-door. Charlie soon discovers his suave new neighbor, Jerry Dandrige, has fangs. When the police don’t believe him, Charlie appeals to the local creature-feature host, washed up horror icon Peter Vincent, for help. &lt;a class=&quot;!no-underline uppercase whitespace-nowrap text-accent text-sm leading-none&quot; href=&quot;/reviews/fright-night-by-john-skipp-craig-spector/&quot;&gt;Continue reading...&lt;/a&gt;&lt;/p&gt;</content:encoded></item></channel></rss>